File: tzio-core/clj/src/clj/refs/references.clj
(ns clj.refs.references
  (:import 
    (com.kineolyan.tzio.v1.api.ref 
      ValueReference
      SlotReference)))

(defn acc-slot
  [_] 
  [:acc])
(def to-acc-slot acc-slot)

(defn nil-slot
  ([_] [:nil]))
(def to-nil-slot nil-slot)

(defn value-slot
  [value]
  [:value value])
(defn to-value-slot
  [^ValueReference type] 
  (value-slot (.-value type)))

(defn ref-slot
  [ref]
  [:slot ref])
(defn to-ref-slot
  [^SlotReference type]
  (ref-slot (.-value type)))

(defn convert-input
  [type]
  (acc-slot))

File: tzio-core/jdk18/src/main/java/com/kineolyan/tzio/v1/java/ref/References.java
package com.kineolyan.tzio.v1.java.ref;

/**
 * Public facade to create references to inputs and outputs.
 */
public class References {

	private References() {}

	public static InputReference inSlot(final int idx) {
		return SlotReference.of(idx);
	}

	public static OutputReference outSlot(final int idx) {
		return SlotReference.of(idx);
	}

	public static AccReference acc() {
		return AccReference.INSTANCE;
	}

	public static InputReference value(final int value) {
		return ValueReference.of(value);
	}

	public static InputReference inNil() {
		return NilReference.INSTANCE;
	}

	public static OutputReference outNil() {
		return NilReference.INSTANCE;
	}

}
